### ISSUE TITLE
Portable Mode (Windows)

### ISSUE THREAD

**Issue by smaragdus** [Taiko2k/TauonMusicBox#79]
Currently (as of version [4.0.0](https://github.com/Taiko2k/TauonMusicBox/releases/tag/v4.0.0)) **Tauon Music Box** saves data and settings (under Windows) in user profile folder:

`C:\Users\User\Music\TauonMusicBox`

I would highly appreciate if there is a way to save program data and settings inside program folder (next to the executable- **tauon.exe**).

Portable mode may be triggered in a variety of ways, I will provide several examples bellow:

- **command-line argument** - if Tauon Music Box is started with a specific parameter (for example- `tauon.exe -portable`) program data and settings are not saved in user profile folder but in program folder;
- **portable flag** - on program start Tauon Music Box checks program folder for portable flag, for example:
`portable` (without extension) or `portable.txt`
and if such file is present data and settings are saved not in user profile folder but in in program folder;
- **empty sub-folder** - on program start Tauon Music Box checks program folder for a specific sub-folder, for example:
data, settings, configuration, TauonMusicBox, etc
and if such sub-folder is present data and settings are saved not in user profile folder but in this sub-folder inside program folder;

Additionally you may add new installation method- Extraction (Portable Mode) which just extracts the installer and does not create uninstaller. Another option might be to distribute the program as archives too which do not require installation- just extract the archive and run the executable.

Any method to use **Tauon Music Box** in portable mode would be fine for me but I prefer options 2 and 3 (_portable flag_ and _empty sub-folder_).

For me portable mode will add some more flexibility to Windows build of **Tauon Music Box**- it would be easier to transfer the program from machine to machine and it will also be easier to backup the program. More, there are people (like me) who prefer portable programs which save data and settings locally- for example if for some reason I need to re-install Windows it is easier to restore all my portable programs (all of them are in a special folder).

**Comment by Taiko2k** [comment]
Actually, if you copy/move the install folder to somewhere outside of the program files folder (So, the text `Program Files` is not in the path you are running from, also, delete `unins000.exe` from the folder), then it will run in portable mode, saving all data to that same folder.

Although, I cannot guarantee no traces would be left in the registry (I don't know if it would or not)

Perhaps in future releases I'll just supply the files without the installer.

**Comment by smaragdus** [comment]
@Taiko2k 

Thank you for the tip. Before opening this issue I browsed through the [Wiki](https://github.com/Taiko2k/TauonMusicBox/wiki) but I did not find this documented. I followed the procedure you suggested and it really resulted in **Tauon Music Box** running in portable mode. This is good enough for me. I wonder- wouldn't it be better if configuration files (star.p.backup3, crash.log, star.p, state.p, config.txt) would be placed in a sub-folder inside program folder- what do you think? Before posting this issue I also checked Windows registry and to my knowledge **Tauon Music Box** does not write to Windows registry at all. Is that true?

In conclusion- I am satisfied- copying the installation folder to non-system folder really triggers portable mode so you may close this issue.

**Comment by smaragdus** [comment]
I forgot to mention that I discovered the installer can be extracted using for example [Universal Extractor 2](https://github.com/Bioruebe/UniExtract2) so upon update one can avoid new installation.

**Comment by Taiko2k** [comment]
Indeed, the installer is not required, I could simply supply the files in a .zip archive.

When upgrading, you'd need to copy those user data/config files to the new folder (Maybe you're right, having these in a sub-folder would make selecting and copying them easier)

As for the windows registry, I don't use it at all (Apart form the installer), except, I don't know if a 3rd party library used writes to it (like the python interpreter), which is why I said I did not know for sure, but good to hear if it does not.

**Comment by smaragdus** [comment]
One last remark upon this topic- I suppose that it would be better if you choose to release the program both as installer and as archive. I have the impression that not so advanced users are used to to 'click next' installers which also create desktop icon. Archives often put such users at bay- some of them have the habit to run the executable directly from the archive without extracting it. I remember that once an acquaintance of mine complained that a portable program crashed. She applied a screen-shot and I could see that the executable was started from within WinRAR and the program (Tox client) could not save its data files. So, if it is not too much of an effort- better provide both installers and archives- thus you will both spare the less advanced users some troubles and yourself- possible ridiculous complaints.

**Comment by smaragdus** [comment]
I forgot to say that I cannot totally guarantee that Tauon Music Box doe not touch the registry at all- the program itself does not but some components might do that. At least I did not find such.

**Comment by smaragdus** [comment]
@Taiko2k 

It is nice that you released separate installer and portable versions. I also think it is tidier that settings are stored in `user-data` sub-folder now.

Regards
